File: rpc/rust/caravan-rpc/src/resources/auto_register.rs
//! Auto-registration: read env vars, call `provide()` for each
//! Caravan-shipped resource seam. Mirrors `caravan_rpc.resources.auto_register`
//! in the Python SDK.
//!
//! Env-var routing:
//! - BlobStore (marker-driven so the caller's compose / HCL emit
//!   declares intent unambiguously; defeats the silent-fallback footgun):
//!   - `CARAVAN_BLOB_BACKEND=s3` → S3BlobStore via
//!     `S3BlobStore::from_env()` (consumes `S3_BUCKET` +
//!     optional `S3_ENDPOINT_URL` / `AWS_*`). Loud-fails when
//!     `S3_BUCKET` is unset.
//!   - `CARAVAN_BLOB_BACKEND=local-fs` → LocalFsBlobStore at
//!     yaml_fallback's `blob_storage.base_path` (or `/data/blobs`
//!     default). Skips S3 even when `S3_ENDPOINT_URL` is set —
//!     the "MinIO emitted but skipped" oss-local case.
//!   - unset → consult `yaml_fallback` (non-caravan local-dev).
//! - MessageQueue (URL-scheme routing — no marker needed since the
//!   scheme is unambiguous):
//!   - `QUEUE_URL` scheme `redis(s)://` → RedisStreamQueue.
//!   - `QUEUE_URL` scheme `amqp(s)://` → RabbitMQQueue.
//!   - `QUEUE_URL` scheme `http(s)://` → SqsQueue.
//!   - none → YAML fallback (best-effort).
//!
//! `yaml_fallback` is the user's parsed application config dict. We
//! consult known sub-keys (`blob_storage.type`, `queue.type`, etc.) but
//! ignore everything else.

use std::sync::Arc;

use serde_yaml::Value;

use super::blob_store::{BlobStore, LocalFsBlobStore};
#[cfg(any(
    feature = "resources-aws",
    feature = "resources-redis",
    feature = "resources-rabbit"
))]
use super::queue::MessageQueue;
use crate::provide;

/// One-line resource bootstrap. Call once at process startup.
/// `yaml_fallback` is optional — pass `None` if your app has no
/// non-Caravan local-dev path.
pub fn auto_register_resources(yaml_fallback: Option<&Value>) -> Result<(), AutoRegisterError> {
    register_blob_store(yaml_fallback)?;
    register_message_queue(yaml_fallback)?;
    Ok(())
}

#[derive(Debug, thiserror::Error)]
pub enum AutoRegisterError {
    #[error("BlobStore registration failed: {0}")]
    BlobStore(String),
    #[error("MessageQueue registration failed: {0}")]
    Queue(String),
    #[error("unsupported QUEUE_URL scheme {0:?}")]
    UnsupportedQueueScheme(String),
}

fn register_blob_store(fallback: Option<&Value>) -> Result<(), AutoRegisterError> {
    // Caravan-emitted explicit marker: `CARAVAN_BLOB_BACKEND` declares
    // which impl the caller's compose/HCL emit intends. With backend=s3
    // we assert S3_BUCKET is set — catches the "user forgot .env.hybrid
    // from tofu output" footgun loudly at startup. With backend=local-fs
    // we skip S3 entirely (oss-local "MinIO emitted but skipped"). No
    // marker → consult yaml_fallback (non-caravan local-dev path).
    match std::env::var("CARAVAN_BLOB_BACKEND").ok().as_deref() {
        Some("s3") => {
            if std::env::var("S3_BUCKET").is_err() {
                return Err(AutoRegisterError::BlobStore(
                    "CARAVAN_BLOB_BACKEND=s3 but S3_BUCKET is unset; did you forget \
                     to populate .env.hybrid from `tofu output -json`? \
                     (`caravan up` auto-generates it.)"
                        .into(),
                ));
            }
            #[cfg(feature = "resources-aws")]
            {
                let s3 = super::blob_store::S3BlobStore::from_env()
                    .map_err(|e| AutoRegisterError::BlobStore(e.to_string()))?;
                provide::<dyn BlobStore>(Arc::new(s3));
                return Ok(());
            }
            #[cfg(not(feature = "resources-aws"))]
            {
                return Err(AutoRegisterError::BlobStore(
                    "CARAVAN_BLOB_BACKEND=s3 but caravan-rpc built without `resources-aws` feature".into(),
                ));
            }
        }
        Some("local-fs") => {
            let base = yaml_blob_base(fallback).unwrap_or_else(|| "/data/blobs".into());
            let store = LocalFsBlobStore::new(&base)
                .map_err(|e| AutoRegisterError::BlobStore(e.to_string()))?;
            provide::<dyn BlobStore>(Arc::new(store));
            return Ok(());
        }
        _ => {}
    }

    // YAML fallback.
    let Some(fb) = fallback else {
        return Ok(());
    };
    let Some(blob) = fb.get("blob_storage") else {
        return Ok(());
    };
    let Some(storage_type) = blob.get("type").and_then(Value::as_str) else {
        return Ok(());
    };
    match storage_type {
        "local_fs" => {
            if let Some(base) = blob.get("base_path").and_then(Value::as_str) {
                let store = LocalFsBlobStore::new(base)
                    .map_err(|e| AutoRegisterError::BlobStore(e.to_string()))?;
                provide::<dyn BlobStore>(Arc::new(store));
            }
        }
        "s3" => {
            #[cfg(feature = "resources-aws")]
            {
                let bucket = blob.get("bucket").and_then(Value::as_str).ok_or_else(|| {
                    AutoRegisterError::BlobStore("blob_storage.s3 requires bucket".into())
                })?;
                let region = blob.get("region").and_then(Value::as_str);
                let store = super::blob_store::S3BlobStore::new(bucket, None, None, None, region)
                    .map_err(|e| AutoRegisterError::BlobStore(e.to_string()))?;
                provide::<dyn BlobStore>(Arc::new(store));
            }
            #[cfg(not(feature = "resources-aws"))]
            return Err(AutoRegisterError::BlobStore(
                "blob_storage.s3 requires `resources-aws` feature".into(),
            ));
        }
        _ => {}
    }
    Ok(())
}

fn register_message_queue(fallback: Option<&Value>) -> Result<(), AutoRegisterError> {
    if let Ok(env_url) = std::env::var("QUEUE_URL") {
        // Parse the scheme without depending on the `url` crate — that's
        // a feature-gated optional dep. Scheme is `<chars>://` so a simple
        // split suffices for our routing decision.
        let scheme = env_url.split("://").next().unwrap_or("").to_lowercase();
        match scheme.as_str() {
            "redis" | "rediss" => {
                #[cfg(feature = "resources-redis")]
                {
                    let consumer_group = consumer_group_from_fallback(fallback);
                    let q = super::queue::RedisStreamQueue::new(&env_url, &consumer_group)
                        .map_err(|e| AutoRegisterError::Queue(e.to_string()))?;
                    provide::<dyn MessageQueue>(Arc::new(q));
                    return Ok(());
                }
                #[cfg(not(feature = "resources-redis"))]
                return Err(AutoRegisterError::Queue(
                    "QUEUE_URL redis:// scheme requires `resources-redis` feature".into(),
                ));
            }
            "amqp" | "amqps" => {
                #[cfg(feature = "resources-rabbit")]
                {
                    let q = super::queue::RabbitMQQueue::new(&env_url)
                        .map_err(|e| AutoRegisterError::Queue(e.to_string()))?;
                    provide::<dyn MessageQueue>(Arc::new(q));
                    return Ok(());
                }
                #[cfg(not(feature = "resources-rabbit"))]
                return Err(AutoRegisterError::Queue(
                    "QUEUE_URL amqp:// scheme requires `resources-rabbit` feature".into(),
                ));
            }
            "http" | "https" => {
                #[cfg(feature = "resources-aws")]
                {
                    let q = super::queue::SqsQueue::from_url(&env_url)
                        .map_err(|e| AutoRegisterError::Queue(e.to_string()))?;
                    provide::<dyn MessageQueue>(Arc::new(q));
                    return Ok(());
                }
                #[cfg(not(feature = "resources-aws"))]
                return Err(AutoRegisterError::Queue(
                    "QUEUE_URL https:// (SQS) scheme requires `resources-aws` feature".into(),
                ));
            }
            other => {
                return Err(AutoRegisterError::UnsupportedQueueScheme(other.to_string()));
            }
        }
    }
    // YAML fallback intentionally minimal — most Rust services that want
    // a no-Caravan local-dev queue path can set QUEUE_URL=redis://localhost
    // in their .env. Extend here when a user repo asks for it.
    let _ = fallback;
    Ok(())
}

#[cfg(feature = "resources-redis")]
fn consumer_group_from_fallback(fallback: Option<&Value>) -> String {
    let default = "caravan-workers".to_string();
    let Some(fb) = fallback else {
        return default;
    };
    fb.get("queue")
        .and_then(|q| q.get("consumer_group"))
        .and_then(Value::as_str)
        .map(String::from)
        .unwrap_or(default)
}

/// Pulls `blob_storage.base_path` from the user's YAML fallback config
/// when type=local_fs. Returns None otherwise. Used by the
/// CARAVAN_BLOB_BACKEND=local-fs path to find the on-disk root.
fn yaml_blob_base(fallback: Option<&Value>) -> Option<String> {
    let fb = fallback?;
    let blob = fb.get("blob_storage")?;
    if blob.get("type").and_then(Value::as_str)? != "local_fs" {
        return None;
    }
    blob.get("base_path")
        .and_then(Value::as_str)
        .map(String::from)
}

File: rpc/python/src/caravan_rpc/resources/auto_register.py
"""auto_register_resources — one-line resource bootstrap.

A user's process startup (worker entry, CLI main()) calls
``auto_register_resources()`` once. caravan-rpc reads its own env vars
(injected by ``caravan compile``), picks the right impl per seam, and calls
``provide()`` so subsequent ``client(BlobStore)`` / ``client(MessageQueue)``
dispatches resolve.

Env-var routing (matches what the caravan compiler emits in
``internal/compiler/resource_endpoints.go``):

  BlobStore:
    S3_ENDPOINT_URL set        → S3BlobStore against MinIO (oss-local)
    S3_BUCKET set, no endpoint → S3BlobStore against real AWS (cloud-managed)
    neither                    → LocalFsBlobStore from yaml_fallback hint
                                  (non-Caravan local-dev path); skipped if
                                  the hint doesn't describe a blob store

  MessageQueue:
    QUEUE_URL scheme:
      redis://    → RedisStreamQueue
      amqp://     → RabbitMQQueue
      https://    → SqsQueue (cloud-managed)
    none        → YAML-fallback dispatch; skipped if hint absent

The ``yaml_fallback`` dict is the user's parsed application config, passed
verbatim. caravan-rpc reads a known sub-shape:

    {
        "blob_storage": {"type": "local_fs", "base_path": "..."} | {"type": "s3", "bucket": "...", "region": "..."},
        "queue":        {"type": "redis_stream", "url": "..."} | {"type": "rabbitmq", "url": "..."},
    }

Unrecognized keys are ignored. Fallback is best-effort — when neither env
vars nor fallback yield a usable impl, the seam stays unregistered and the
first ``client(BlobStore).put(...)`` raises ``LookupError`` with a clear
message.
"""

from __future__ import annotations

import os
from typing import Any
from urllib.parse import urlparse

from .. import provide
from .blob_store import BlobStore, LocalFsBlobStore, S3BlobStore
from .queue import MessageQueue, RabbitMQQueue, RedisStreamQueue, SqsQueue


def auto_register_resources(yaml_fallback: dict | None = None) -> None:
    """Read env vars + optional yaml fallback; call ``provide()`` for each
    resource seam caravan-rpc can satisfy. Idempotent within one process —
    re-registering replaces the prior impl (last-write-wins, matches
    ``provide()``'s own semantics).
    """
    _register_blob_store(yaml_fallback)
    _register_message_queue(yaml_fallback)


def _register_blob_store(fallback: dict | None) -> None:
    # Caravan-emitted explicit marker: `CARAVAN_BLOB_BACKEND` declares
    # which impl the caller's compose/HCL emit intends. With backend=s3
    # the SDK asserts S3_BUCKET is set — catches the "user forgot
    # .env.hybrid from tofu output" footgun loudly at startup instead of
    # silently masquerading cloud-managed as LocalFs. With backend=local-fs
    # the SDK skips S3 entirely (oss-local "MinIO emitted but skipped").
    # No marker → fall through to yaml_fallback (non-caravan local dev).
    backend = os.environ.get("CARAVAN_BLOB_BACKEND")
    if backend == "s3":
        if not os.environ.get("S3_BUCKET"):
            raise ValueError(
                "CARAVAN_BLOB_BACKEND=s3 but S3_BUCKET is unset; "
                "did you forget to populate .env.hybrid from `tofu output -json`? "
                "(`caravan up` auto-generates it.)"
            )
        provide(BlobStore, S3BlobStore.from_env())
        return
    if backend == "local-fs":
        base = _yaml_blob_base(fallback) or "/data/blobs"
        provide(BlobStore, LocalFsBlobStore(base))
        return

    # No marker — non-caravan local dev path; consult yaml_fallback only.
    if not fallback:
        return
    blob = fallback.get("blob_storage")
    if not isinstance(blob, dict):
        return
    storage_type = blob.get("type")
    if storage_type == "local_fs":
        base = blob.get("base_path")
        if base:
            provide(BlobStore, LocalFsBlobStore(base))
    elif storage_type == "s3":
        bucket = blob.get("bucket")
        if not bucket:
            return
        provide(
            BlobStore,
            S3BlobStore(bucket=bucket, region=blob.get("region")),
        )


def _yaml_blob_base(fallback: dict | None) -> str | None:
    """Pull `blob_storage.base_path` from the user's YAML fallback config,
    if present and the type is local_fs. Returns None otherwise.
    """
    if not fallback:
        return None
    blob = fallback.get("blob_storage")
    if not isinstance(blob, dict):
        return None
    if blob.get("type") != "local_fs":
        return None
    base = blob.get("base_path")
    return base if isinstance(base, str) and base else None


def _register_message_queue(fallback: dict | None) -> None:
    env_url = os.environ.get("QUEUE_URL", "")
    if env_url:
        scheme = urlparse(env_url).scheme
        consumer_group = _consumer_group_from_fallback(fallback)
        impl: Any
        if scheme in ("redis", "rediss"):
            impl = RedisStreamQueue(env_url, consumer_group)
        elif scheme in ("amqp", "amqps"):
            impl = RabbitMQQueue.from_url(env_url)
        elif scheme in ("http", "https"):
            impl = SqsQueue.from_url(env_url)
        else:
            raise ValueError(f"QUEUE_URL scheme {scheme!r} not supported")
        provide(MessageQueue, impl)
        return

    # YAML fallback.
    if not fallback:
        return
    queue = fallback.get("queue")
    if not isinstance(queue, dict):
        return
    qtype = queue.get("type")
    consumer_group = queue.get("consumer_group", "caravan-workers")
    if qtype == "redis_stream":
        url = queue.get("url")
        if url:
            provide(MessageQueue, RedisStreamQueue(url, consumer_group))
    elif qtype == "rabbitmq":
        url = queue.get("url")
        if url:
            provide(MessageQueue, RabbitMQQueue(url))
    elif qtype == "sqs":
        # SQS via YAML hint: requires QUEUE_URL env (set by Caravan from
        # tofu output). Documented; no silent fallback.
        raise ValueError(
            "queue.type=sqs requires QUEUE_URL env var (caravan emits it from tofu output)"
        )


def _consumer_group_from_fallback(fallback: dict | None) -> str:
    if not fallback:
        return "caravan-workers"
    queue = fallback.get("queue")
    if isinstance(queue, dict):
        return queue.get("consumer_group", "caravan-workers")
    return "caravan-workers"
